### ISSUE TITLE
Block editor: Focus loss when switching Visual editor / Code editor

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#52719]
### Description

Splitting this out from https://github.com/WordPress/gutenberg/issues/52301#issuecomment-1635713097

When switching between Visual editor / Code editor and back, a keyboard focus loss occurs. There are two problems with this:
1. For keyboard users and assistive technology users a focus loss is a terrible experience as they are 'catapulted' back to the document root and they have to start navigating the page from scratch.
2. When a full focus loss occurs, all the Editor keyboard shortcuts don't work any longer. For example Cmd+K to open the command center doesn't do anything. See https://github.com/WordPress/gutenberg/issues/52301

Switching between Visual editor / Code editor should implement some kind of focus management to make sure focus stays in the most appropriate place but I'd think there are different cases to take into consideration.

- Switching via the Options menu regardless if using a pointing device (e.g. mouse) or keyboard: I think focus should stay on the menu item button that was activated.
- Instead, when switching via keyboard shortcut Shift+Option+Command+M focus should stay where it was, or be set to the most appropriate place e.g.:
  - One of the blocks is selected? Shift+Option+Command+M > focus is set on the Code textarea > Shift+Option+Command+M > focus is set back on the block that was selected.
  - Any other UI control is selected? Shift+Option+Command+M > focus stays on the focused control, whatever it is.
  - I may be missing other scenarios but in general, focus should never be lost and should be set to the most appropriate place.

Implementing some proper focus management would also solve the editor keyboard shortcuts issue (e.g. Cmd+K).

Anecdotally: I often use keyboard shortcuts when available. I found myself often in the need of switching the editor to Code editor and back to Visual. For me, the faster way to do so is by pressing `Shift+Option+Command+M` and it's a bit annoying that once the editor switches to Code editor I can't use `Shift+Option+Command+M` to switch it back to Visual _unless I click within the Code editor first_, which defeats the purpose of using keyboard shortcuts. 

### Step-by-step reproduction instructions

- Go to the post editor or site editor.
- Select any block.
- Press the `Shift+Option+Command+M` keys.
- The editor switches to Code editor.
- Observe that focus is lost:
  - No focus style indication on any element.
  - Pressing `Shift+Option+Command+M` again does not switch back the editor to VIsual editor.
  - Any other keyboard shortcut doesn't work e.g.:
  - Control+Cmd+H doesn't do anything: it should open the keyboard shortcuts modal dialog.
  - Cmd+K doesn't do anything: it should open the Command center modal dialog.
- Click with your mouse within the Code editor area.
- Observe it gets focused and a blue outline focus style is now visible.
- Try again the keyboard shortcuts and observe they now work.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by Mamaduka** [comment]
Probably a duplicate of #42145, and we can close #40577 once this is resolved.

I need to double-check this, but I think the focus is lost when the visual editor has it and we're switching to a code editor via keyboard shortcuts. This happens because one editor needs to unmount to re-render the second.

Unfortunately, we can't focus on a currently selected block, as the code editor has no notion of the blocks; it just renders a text field and a string.

If you have any suggestions on "the best way" to solve this, I'll be happy to work on it.

**Comment by afercia** [comment]
> Unfortunately, we can't focus on a currently selected block, as the code editor has no notion of the blocks; it just renders a text field and a string.

Thanks @Mamaduka Maybe we can 'remember' what the selected block was before switching to Code editor and set focus to it again when switching back to Visual? Maybe the most tricky part is that all this should happen only when using the keyboard shortcut Shift+Option+Command+M.

**Comment by afercia** [comment]
I'll close in favor of https://github.com/WordPress/gutenberg/issues/42145, and add the accessibility label and some more context there.

**Comment by Mamaduka** [comment]
Thank you, let's continue our discussion there.
